src/dynamic_contract.rs

```diff
@@ -0,0 +1,280 @@
+extern crate bincode;
+extern crate generic_array;
+
+use bank::{Account, Result};
+use bincode::deserialize;
+use libloading::{Library, Symbol};
+use signature::Pubkey;
+use std::path::PathBuf;
+use transaction::Transaction;
+
+#[cfg(debug_assertions)]
+const CARGO_PROFILE: &str = "debug";
+
+#[cfg(not(debug_assertions))]
+const CARGO_PROFILE: &str = "release";
+
+/// Dynamic link library prefix
+#[cfg(unix)]
+const PLATFORM_FILE_PREFIX: &str = "lib";
+/// Dynamic link library prefix
+#[cfg(windows)]
+const PLATFORM_FILE_PREFIX: &str = "";
+/// Dynamic link library file extension specific to the platform
+#[cfg(any(target_os = "macos", target_os = "ios"))]
+const PLATFORM_FILE_EXTENSION: &str = "dylib";
+/// Dynamic link library file extension specific to the platform
+#[cfg(all(unix, not(any(target_os = "macos", target_os = "ios"))))]
+const PLATFORM_FILE_EXTENSION: &str = "so";
+/// Dynamic link library file extension specific to the platform
+#[cfg(windows)]
+const PLATFORM_FILE_EXTENSION: &str = "dll";
+
+/// Creates a platform-specific file path
+fn create_library_path(name: &str) -> PathBuf {
+    let mut path = PathBuf::new();
+    path.push("target");
+    path.push(CARGO_PROFILE);
+    path.push("deps");
+    path.push(PLATFORM_FILE_PREFIX.to_string() + name);
+    path.set_extension(PLATFORM_FILE_EXTENSION);
+    path
+}
+
+// All contracts export a symbol named process()
+const ENTRYPOINT: &str = "process";
+
+type Entrypoint = unsafe extern "C" fn(tx: &Transaction, accounts: &mut [Account]) -> Result<()>;
+
+pub const DYNAMIC_CONTRACT_ID: [u8; 32] = [
+    2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+];
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum DynamicContract {
+    /// Native contract
+    /// * Transaction::keys[0..] - contract dependent
+    /// * name - name of the contract, translated to a file path of the contract module
+    /// * userdata - contract specific user data
+    Native { name: String, userdata: Vec<u8> },
+    /// BPF contract
+    /// * Transaction::keys[0..] - contract dependent
+    /// * TODO BPF specific stuff
+    /// * userdata - contract specific user data
+    BPF { userdata: Vec<u8> },
+}
+
+impl DynamicContract {
+    pub fn id() -> Pubkey {
+        Pubkey::new(&DYNAMIC_CONTRACT_ID)
+    }
+
+    pub fn check_id(contract_id: &Pubkey) -> bool {
+        contract_id.as_ref() == DYNAMIC_CONTRACT_ID
+    }
+
+    pub fn get_balance(account: &Account) -> i64 {
+        account.tokens
+    }
+
+    pub fn process_transaction(tx: &Transaction, accounts: &mut [Account]) {
```

Comment:
Two or more different transactions involved.  First one to load the contract (tx usedata contains filename and id) and subsequent ones to call the loaded contract by id (tx contains contract userdata only).  Did you have thoughts on unloading, a second new instruction to `SystemContract`?

Also, are you thinking that `budget_contract` would also be dynamic?

Reply:
> Two or more different transactions involved...
Yep, exactly.

> Did you have thoughts on unloading...
For now let's punt on unloading.  But I think unloading can be implicit -- once there are no active Accounts with the `contract_id` for the given .so in bank.rs, that .so can be unloaded from memory.  It would be reloaded again once a new active Account is tagged with that `contract_id`.  Unloading will probably be associated with a token value as well.  The user pays 100 (N) tokens to make that `contract_id` available for 100 (M) days.

> Also, are you thinking that budget_contract would also be dynamic?

It certainly could be eventually.  There's a bunch of code detangling that needs to happen in transaction.rs/bank.rs/.. before that'll be possible though.

Reply:
Sounds good, will give it a shot

Reply:
@mvines, would you like me to merge this PR and then start a new one for the structure changes you suggested?

Also, I'll make the changes to pass the keys in this PR, agreed?

Reply:
There’s no rush to land this so I’d rather the structural changes are made in this PR (keys too)